control-plane-operator/controllers/hostedcontrolplane/cloud/kubevirt/reconcile.go

```diff
@@ -0,0 +1,199 @@
+	ownerRef.ApplyTo(role)
+	role.Rules = []rbacv1.PolicyRule{
+		{
+			APIGroups: []string{"kubevirt.io"},
+			Resources: []string{"virtualmachines"},
+			Verbs: []string{
+				"get",
+				"list",
+				"watch",
+			},
+		},
+		{
+			APIGroups: []string{"kubevirt.io"},
+			Resources: []string{"virtualmachineinstances"},
+			Verbs: []string{
+				"get",
+				"list",
+				"watch",
+				"update",
+			},
+		},
+		{
+			APIGroups: []string{""},
+			Resources: []string{"services"},
+			Verbs:     []string{rbacv1.VerbAll},
+		},
+	}
+	return nil
+}
+
+func ReconcileCCMRoleBinding(roleBinding *rbacv1.RoleBinding, ownerRef config.OwnerRef, sa *corev1.ServiceAccount, role *rbacv1.Role) error {
+	ownerRef.ApplyTo(roleBinding)
+	roleBinding.RoleRef = rbacv1.RoleRef{
+		APIGroup: rbacv1.GroupName,
+		Kind:     "Role",
+		Name:     role.Name,
+	}
+	roleBinding.Subjects = []rbacv1.Subject{
+		{
+			Namespace: sa.Namespace,
+			Kind:      rbacv1.ServiceAccountKind,
+			Name:      sa.Name,
+		},
+	}
+	return nil
+}
+
+func ReconcileDeployment(deployment *appsv1.Deployment, hcp *hyperv1.HostedControlPlane, serviceAccountName string, releaseImage *releaseinfo.ReleaseImage) error {
+	clusterName, ok := hcp.Labels["cluster.x-k8s.io/cluster-name"]
+	if !ok {
+		return fmt.Errorf("\"cluster.x-k8s.io/cluster-name\" label doesn't exist in HostedControlPlane")
+	}
+	deploymentConfig := newDeploymentConfig()
+	deployment.Spec = appsv1.DeploymentSpec{
+		Selector: &metav1.LabelSelector{
+			MatchLabels: ccmLabels(),
+		},
+		Strategy: appsv1.DeploymentStrategy{
+			Type: appsv1.RecreateDeploymentStrategyType,
+		},
+		Template: corev1.PodTemplateSpec{
+			ObjectMeta: metav1.ObjectMeta{
+				Labels: ccmLabels(),
+			},
+			Spec: corev1.PodSpec{
+				Containers: []corev1.Container{
+					util.BuildContainer(CCMContainer(), buildCCMContainer(clusterName, releaseImage)),
+				},
+				Volumes:            []corev1.Volume{},
+				ServiceAccountName: serviceAccountName,
+			},
+		},
+	}
+
+	addVolumes(deployment)
+
+	config.OwnerRefFrom(hcp).ApplyTo(deployment)
+	deploymentConfig.ApplyTo(deployment)
+	return nil
+}
+
+func addVolumes(deployment *appsv1.Deployment) {
+
+	deployment.Spec.Template.Spec.Volumes = append(
+		deployment.Spec.Template.Spec.Volumes,
+		util.BuildVolume(ccmVolumeKubeconfig(), buildCCMVolumeKubeconfig),
+	)
+	deployment.Spec.Template.Spec.Volumes = append(
+		deployment.Spec.Template.Spec.Volumes,
+		util.BuildVolume(ccmCloudConfig(), buildCCMCloudConfig),
+	)
+}
+
+func podVolumeMounts() util.PodVolumeMounts {
+	return util.PodVolumeMounts{
+		CCMContainer().Name: util.ContainerVolumeMounts{
+			ccmVolumeKubeconfig().Name: "/etc/kubernetes/kubeconfig",
+			ccmCloudConfig().Name:      "/etc/cloud",
+		},
+	}
+}
+
+func buildCCMContainer(clusterName string, releaseImage *releaseinfo.ReleaseImage) func(c *corev1.Container) {
+	return func(c *corev1.Container) {
+		c.Image = releaseImage.ComponentImages()["kubevirt-cloud-controller-manager"]
```

Comment:
what should we do if the `kubevirt-cloud-controller-manager` image doesn't exist in the payload because someone is using an older release payload? Should that situation be handled?

Reply:
I think that this is going to be supported only with 4.12 release image, so `kubevirt-cloud-controller-manager` is going to be included for sure
But better have input about it from someone in hypershift team

Reply:
you're right. I keep forgetting that the CPO itself is a part of the release payload. So this code will only be included in a release payload that has the kccm.